test/click_cli/test_cmain.py

```diff
@@ -0,0 +1,499 @@
+import os
+import sys
+
+import pytest
+import yaml
+from mock import call, patch
+
+from nameko.click_cli import cli
+from nameko.click_cli.utils import import_services
+from nameko.click_cli.utils.config import ENV_VAR_MATCHER, setup_yaml_parser
+from nameko.exceptions import CommandError, ConfigurationError
+
+
+try:
+    import regex
+except ImportError:  # pragma: no cover
+    has_regex_module = False
+else:  # pragma: no cover
+    del regex
+    has_regex_module = True
+
+
+@pytest.yield_fixture(autouse=True)
+def fake_argv(empty_config):
+    with patch.object(
+        sys,
+        "argv",
+        [
+            "cnameko",
+            "run",
+            "--define",
+            "AMQP_URI=pyamqp://someuser:*****@somehost/",
+            "test.sample:Service",
+        ],
+    ):
+        yield
+
+
+class ExpectedCall(object):
+    """Class to generate calls expected by mocked functions
+    """
+
+    def __init__(
+        self,
+        config_file=None,
+        define=None,
+        services=None,
+        broker=None,
+        backdoor_port=None,
+        interface=None,
+    ):
+        self.config_file = config_file
+        self.define = define
+        self.services = services
+        self.broker = broker
+        self.backdoor_port = backdoor_port
+
+    @property
+    def setup_config_call(self):
+        return call(self.config_file, self.define, self.broker)
+
+    @property
+    def main_run_call(self):
+        return call(self.services, self.backdoor_port)
+
+
+@pytest.fixture()
+def expected():
+    define = {"AMQP_URI": "pyamqp://someuser:*****@somehost/"}
+    services = import_services("test.sample:Service")
+    return ExpectedCall(define=define, services=services)
+
+
+def test_run(expected):
+    with patch("nameko.click_cli.run.main") as main_run:
+        with patch("nameko.click_cli.setup_config") as setup_config:
+            cli(standalone_mode=False)
+            assert setup_config.call_count == 1
+            assert main_run.call_count == 1
+            assert setup_config.call_args == expected.setup_config_call
+            assert main_run.call_args == expected.main_run_call
```

Comment:
When I want to assert exactly one call _and_ check the arguments I usually use `assert call_args_list == [call(<arguments>)]` May help make this a bit more terse.

Reply:
Great trick. `nameko` test suite is one big lesson on testing with mockups for me.
I will change that.